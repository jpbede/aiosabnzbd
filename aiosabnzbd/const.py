"""Constants for aiosabnzbd."""

from enum import StrEnum

API_BASE_URL = "https://api.electricitymap.org/v3"


class ApiEndpoints:
    """Class holding API endpoints."""

    CARBON_INTENSITY_HA = API_BASE_URL + "/home-assistant"
    ZONES = API_BASE_URL + "/zones"
    LATEST_CARBON_INTENSITY = API_BASE_URL + "/carbon-intensity/latest"
    HISTORY_CARBON_INTENSITY = API_BASE_URL + "/carbon-intensity/history"
    LATEST_POWER_BREAKDOWN = API_BASE_URL + "/power-breakdown/latest"
    HISTORY_POWER_BREAKDOWN = API_BASE_URL + "/power-breakdown/history"


class QueueStatus(StrEnum):
    """Enum for queue status."""

    IDLE = "Idle"
    QUEUED = "Queued"
    PAUSED = "Paused"
    DOWNLOADING = "Downloading"
    PROPAGATING = "Propagating"
    FETCHING = "Fetching"


class QueueOperationCommand(StrEnum):
    """Enum for queue operation command."""

    RESUME = "resume"
    PAUSE = "pause"
