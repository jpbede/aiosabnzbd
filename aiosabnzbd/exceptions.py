"""Exceptions for SABnzbd."""


class SabnzbdError(Exception):
    """Generic error occurred in SABnzbd package."""


class SabnzbdConnectionError(SabnzbdError):
    """Error occurred while communicating to the SABnzbd API."""


class SabnzbdConnectionTimeoutError(SabnzbdError):
    """Timeout occurred while connecting to the SABnzbd API."""


class SabnzbdInvalidAPIKeyError(SabnzbdError):
    """Given API Key is invalid."""


class SabnzbdMissingAPIKeyError(SabnzbdError):
    """API Key is missing."""
