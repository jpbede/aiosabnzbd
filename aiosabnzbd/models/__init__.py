"""Models to the SABnzbd API."""

from .base import SabnzbdRequest
from .queue import Queue, QueueResponse, Slot
from .status import StatusResponse

__all__ = [
    "Queue",
    "Slot",
    "QueueResponse",
    "StatusResponse",
    "SabnzbdRequest",
]
