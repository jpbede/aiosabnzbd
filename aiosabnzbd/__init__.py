"""SABnzbd wrapper."""

from .const import QueueOperationCommand, QueueStatus
from .exceptions import (
    SABnzbdConnectionError,
    SABnzbdConnectionTimeoutError,
    SABnzbdError,
    SABnzbdInvalidAPIKeyError,
    SABnzbdMissingAPIKeyError,
)
from .models import History, HistoryResponse
from .models.queue import Queue, QueueResponse, Slot
from .models.status import StatusResponse
from .sabnzbd import SABnzbdClient

__all__ = [
    "History",
    "HistoryResponse",
    "Queue",
    "QueueOperationCommand",
    "QueueResponse",
    "QueueStatus",
    "SABnzbdClient",
    "SABnzbdConnectionError",
    "SABnzbdConnectionTimeoutError",
    "SABnzbdError",
    "SABnzbdInvalidAPIKeyError",
    "SABnzbdMissingAPIKeyError",
    "Slot",
    "StatusResponse",
]
