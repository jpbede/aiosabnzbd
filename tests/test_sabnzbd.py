"""Tests for the SABnzbd client."""

from aioresponses import aioresponses
import pytest
from syrupy.assertion import SnapshotAssertion

from aiosabnzbd import (
    QueueOperationCommand,
    SABnzbdClient,
    SABnzbdConnectionError,
    SABnzbdConnectionTimeoutError,
    SABnzbdInvalidAPIKeyError,
    SABnzbdMissingAPIKeyError,
    StatusResponse,
)

from . import load_fixture


async def test_json_request_without_session(
    snapshot: SnapshotAssertion, responses: aioresponses
) -> None:
    """Test JSON response is handled correctly without given session."""
    responses.get(
        "http://localhost:8080/api?apikey=ab123&mode=queue&output=json",
        body=load_fixture("queue.json"),
    )

    async with SABnzbdClient(url="http://localhost:8080", api_key="ab123") as c:
        assert await c.queue() == snapshot
        assert c.session is not None

    assert c.session.closed


async def test_catch_connection_error(
    client: SABnzbdClient, responses: aioresponses
) -> None:
    """Test JSON response is handled correctly with given session."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=queue",
        status=500,
        headers={"Content-Type": "application/json"},
        body="Boooom!",
    )
    with pytest.raises(SABnzbdConnectionError):
        await client.queue()


async def test_timeout(
    client: SABnzbdClient,
    responses: aioresponses,
) -> None:
    """Test request timeout."""
    responses.add(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=queue",
        timeout=True,
    )
    with pytest.raises(SABnzbdConnectionTimeoutError):
        await client.queue()


async def test_queue(
    client: SABnzbdClient, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test getting the queue."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=queue",
        body=load_fixture("queue.json"),
    )

    assert snapshot == await client.queue()


async def test_operate_queue(client: SABnzbdClient, responses: aioresponses) -> None:
    """Test operating the queue."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=pause",
        body='{"status": true}',
    )

    response = await client.operate_queue(command=QueueOperationCommand.PAUSE)

    assert response == StatusResponse(status=True)


async def test_missing_api_key(client: SABnzbdClient, responses: aioresponses) -> None:
    """Test missing API key."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=queue",
        body="API Key Required",
    )

    with pytest.raises(SABnzbdMissingAPIKeyError):
        await client.queue()


async def test_invalid_api_key(client: SABnzbdClient, responses: aioresponses) -> None:
    """Test invalid API key."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=queue",
        body="API Key Incorrect",
    )

    with pytest.raises(SABnzbdInvalidAPIKeyError):
        await client.queue()


async def test_set_speed_limit(client: SABnzbdClient, responses: aioresponses) -> None:
    """Test setting speed limit."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=config&name=speedlimit&value=50",
        body='{"status": true}',
    )

    response = await client.set_speed_limit(percentage=50)

    assert response == StatusResponse(status=True)


async def test_history(
    client: SABnzbdClient, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test getting the queue."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=history",
        body=load_fixture("history.json"),
    )

    assert snapshot == await client.history()


async def test_version(client: SABnzbdClient, responses: aioresponses) -> None:
    """Test getting the version."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=version",
        body='{"version": "3.0.0"}',
    )

    assert await client.version() == "3.0.0"


async def test_combined_queue_history(
    client: SABnzbdClient, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test getting the queue and history."""
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=queue",
        body=load_fixture("queue.json"),
    )
    responses.get(
        "http://localhost:8080/api?apikey=abc123&output=json&mode=history",
        body=load_fixture("history.json"),
    )

    assert snapshot == await client.combined_queue_history()
