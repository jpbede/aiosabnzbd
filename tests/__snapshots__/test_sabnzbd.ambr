# serializer version: 1
# name: test_history
  History(total_size='1.6 T', month_size='43.0 G', week_size='0 ', day_size='0 ', slots=[], ppslots=0, noofslots=0, last_history_update=4340)
# ---
# name: test_json_request_without_session
  Queue(status=<QueueStatus.PAUSED: 'Paused'>, speedlimit='100', speedlimit_absolut='26214400', paused=True, noofslots_total=0, noofslots=0, limit=0, start=0, timeleft=datetime.timedelta(seconds=6340), speed='0 ', kb_per_sec='0.00', size='0 B', size_left='0 B', megabyte='0.00', megabyte_left='0.00', slots=[], diskspace1='87.81', diskspace2='1815.82', diskspace_total1='116.66', diskspace_total2='3686.62', diskspace1_norm='87.8 G', diskspace2_norm='1.8 T', have_warnings='5', pause_int='0', left_quota='0 ', version='4.2.3', finish=0, cache_article='0', cache_size='0 B', finish_action='None', paused_all=False, quota='0 ', have_quota=False)
# ---
# name: test_queue
  Queue(status=<QueueStatus.PAUSED: 'Paused'>, speedlimit='100', speedlimit_absolut='26214400', paused=True, noofslots_total=0, noofslots=0, limit=0, start=0, timeleft=datetime.timedelta(seconds=6340), speed='0 ', kb_per_sec='0.00', size='0 B', size_left='0 B', megabyte='0.00', megabyte_left='0.00', slots=[], diskspace1='87.81', diskspace2='1815.82', diskspace_total1='116.66', diskspace_total2='3686.62', diskspace1_norm='87.8 G', diskspace2_norm='1.8 T', have_warnings='5', pause_int='0', left_quota='0 ', version='4.2.3', finish=0, cache_article='0', cache_size='0 B', finish_action='None', paused_all=False, quota='0 ', have_quota=False)
# ---
